asu/routers/api.py

```diff
@@ -60,80 +65,48 @@ def validate_request(build_request: BuildRequest):
-            400,
-        )
+        return validation_failure("Handling `defaults` not enabled on server")
 
     if build_request.distro not in get_distros():
-        return (
-            {"detail": f"Unsupported distro: {build_request.distro}", "status": 400},
-            400,
-        )
+        return validation_failure(f"Unsupported distro: {build_request.distro}")
 
     branch = get_branch(build_request.version)["name"]
 
     r = get_redis_client()
 
     if not r.sismember("branches", branch):
-        return (
-            {"detail": f"Unsupported branch: {build_request.version}", "status": 400},
-            400,
-        )
+        return validation_failure(f"Unsupported branch: {build_request.version}")
 
     if not r.sismember(f"versions:{branch}", build_request.version):
-        return (
-            {"detail": f"Unsupported version: {build_request.version}", "status": 400},
-            400,
-        )
+        return validation_failure(f"Unsupported version: {build_request.version}")
 
-    build_request.packages: list[str] = list(
-        map(
-            lambda x: x.removeprefix("+"),
-            (build_request.packages_versions.keys() or build_request.packages),
-        )
-    )
+    build_request.packages: list[str] = [
+        x.removeprefix("+")
+        for x in (build_request.packages_versions.keys() or build_request.packages)
+    ]
 
-    logging.debug("Profile before mapping " + build_request.profile)
+    logging.debug(f"Profile before mapping {build_request.profile = }")
 
     if not r.hexists(f"targets:{branch}", build_request.target):
-        return (
-            {"detail": f"Unsupported target: {build_request.target}", "status": 400},
-            400,
-        )
-
-    if build_request.target in [
-        "x86/64",
-        "x86/generic",
-        "x86/geode",
-        "x86/legacy",
-        "armsr/armv7",
-        "armsr/armv8",
-    ]:
-        logging.debug("Use generic profile for {build_request.target}")
-        build_request.profile = "generic"
+        return validation_failure(f"Unsupported target: {build_request.target}")
+
+    sanitized_profile = build_request.profile.replace(",", "_")
+    target_key = f"{branch}:{build_request.version}:{build_request.target}"
+    profiles_key = f"profiles:{target_key}"
+    if r.sismember(profiles_key, sanitized_profile):
+        logging.debug(f"Using {sanitized_profile = }")
+        build_request.profile = sanitized_profile
     else:
-        if r.sismember(
-            f"profiles:{branch}:{build_request.version}:{build_request.target}",
-            build_request.profile.replace(",", "_"),
-        ):
-            build_request.profile = build_request.profile.replace(",", "_")
+        mapped_profile = r.hget(f"mapping:{target_key}", build_request.profile)
+        logging.debug(f"Checking {mapped_profile = }")
+        if mapped_profile:
+            build_request.profile = mapped_profile
+        elif r.sismember(profiles_key, "generic"):
```

Comment:
What do you think of this approach @dangowrt ? I wonder that if a profile name changes between releases, it would automatically select generic instead, may breaking the running system more than intended.

Reply:
Hmm, yeah that would be bad.  Do we need a profile_changed table like package_changes?  (Maybe 
injects appropriate data into `mapping:`...)

Here are the `generic` profiles I found when I downloaded all of snapshot while doing preliminary work for this PR):
```
target: armsr/armv7          has profiles ['generic']
target: armsr/armv8          has profiles ['generic']
target: ath25/generic        has profiles ['generic', 'ubnt_picostation-2']
target: loongarch64/generic  has profiles ['generic']
target: octeon/generic       has profiles ['cisco_vedge1000', 'generic', 'itus_shield-router', 'ubnt_edgerouter', 'ubnt_edgerouter-4', 'ubnt_edgerouter-6p', 'ubnt_edgerouter-lite', 'ubnt_unifi-usg']
target: x86/64               has profiles ['generic']
target: x86/generic          has profiles ['generic']
target: x86/geode            has profiles ['generic', 'geos']
target: x86/legacy           has profiles ['generic']
```

Wasn't there a target change `mt78??` or something to `filogic` a while back, I wasn't paying attention but it seems to spark something in my memory.  How was that handled?  (As I recall, the `mtxx` name only appeared in snapshot, and then only for a short while???  So had to be done by user???)

I suppose the real solution would be make sure board.json contains a legit profile, make sure x86 and armsr both have `generic` when that's what it really is.  That or list the actual device name (one of my x86's is a `pc-engines-apu2`) in the profiles.json.

Reply:
Looking at that list again, seems like maybe just also check if there's any choice in the matter:
```
    if len(target.profiles) == 1 and target.profiles[0] == "generic":
        # Assume "generic" as that's the only choice possible.
        profile = "generic"
```
I like this one.

Reply:
Changed it to:
```
    elif r.scard(profiles_key) == 1 and r.sismember(profiles_key, "generic"):
```
Errors appropriately on wacky cases (geode, octeon, ath25), selects generic for all of the appropriate cases (x86, armsr, loongarch).